### ISSUE TITLE
CREATE VIEW command denied to user

### ISSUE THREAD

**Issue by smerriman** [AffiliateWP/AffiliateWP#1688]
I'm migrating my site to a new server, and have come across this error message when running the export/import of the database.

CREATE VIEW command denied to user xxx@xxx for table 'xxx_affiliate_wp_campaigns'

I found a past ticket on this issue which just said you needed to elevate permissions to allow this command to be run. However, my host, and likely plenty of other shared hosts, specifically disable the CREATE VIEW command:

https://mediatemple.net/community/products/grid/204405564/database-privileges-on-the-grid#3

Does this mean AffiliateWP cannot be run at all on MediaTemple GS?

**Comment by smerriman** [comment]
(Please don't answer the last question with yes, there must be some workaround possible :) - it's one of the most used hosts.)

**Comment by pippinsplugins** [comment]
Please see this FAQ: http://docs.affiliatewp.com/article/1416-how-to-back-up-affiliate-data

**Comment by smerriman** [comment]
That doesn't appear to help. It tells me I do not need to back the view up, but my issue is that I am completely unable to create it on the new server.

**Comment by smerriman** [comment]
And indeed the note at the end says the CREATE VIEW permission is required. So it seems the plugin is incompatible with Mediatemple after all?

**Comment by pippinsplugins** [comment]
Correct: you do not need to backup the view. It is created automatically
based on the values of the other tables.

**Comment by smerriman** [comment]
Please re-read my message above. It is unable to be automatically created, since the shared host blocks access to CREATE VIEW.

**Comment by pippinsplugins** [comment]
@smerriman I am intimately familiar with the problem.

It only blocks access to create the view when trying to do it during a migration.

Trust me please :)

During the migration / backup, skip the view. AffiliateWP will automatically create it when you load the plugin.

**Comment by smerriman** [comment]
Sorry, I'm obviously not being clear here - MediaTemple GS do not allow *any* sql CREATE VIEW commands to be run, regardless of who is running them.

On the newly migrated site, the view does not appear in phpMyAdmin. Every time I deactivate and reactivate the plugin (ie, the activation hook gets called), I get the following errors logged:

```
mod_fcgid: stderr: WordPress database error CREATE VIEW command denied to user 'xxx'@'xxx' for table 'xxx_affiliate_wp_campaigns' for query CREATE OR REPLACE VIEW xxx_affiliate_wp_campaigns AS
mod_fcgid: stderr: \t\t\t\tSELECT affiliate_id,
mod_fcgid: stderr: \t\t\t\t\tcampaign,
mod_fcgid: stderr: \t\t\t\t\tCOUNT(url) as visits,
mod_fcgid: stderr: \t\t\t\t\tCOUNT(DISTINCT url) as unique_visits,
mod_fcgid: stderr: \t\t\t\t\tSUM(IF(referral_id<>0,1,0)) as referrals,
mod_fcgid: stderr: \t\t\t\t\tROUND((SUM(IF(referral_id<>0,1,0))/COUNT(url)) * 100, 2) as conversion_rate
mod_fcgid: stderr: \t\t\t\tFROM cre8d2015_affiliate_wp_visits GROUP BY affiliate_id, campaign; made by Affiliate_WP_Campaigns_DB->create_view
```

**Comment by pippinsplugins** [comment]
Ah! That is a different story, sorry I thought you meant you were getting the error when trying to export/import the DB.

Ok in that case the _only_ thing (aside from moving off of Media Temple) you can do is go to Media Temple and ask them to grant permission to create views to your SQL user. I would love to have a better answer but that's it.

**Comment by smerriman** [comment]
Hm, OK, that's a pretty significant issue then. I'd suggesting putting somewhere prominent on your website/plugin the fact that it doesn't work on MediaTemple or any other shared host that blocks this (there are several) - and see if you can find a more stable method in the future.

No idea what we'll do in the meantime.

**Comment by smerriman** [comment]
Having looked through the code, am I correct that the dashboard stats tab is *only* place throughout the entire plugin where this view is actually used? And that table appears to not show if the view doesn't exist. So perhaps the plugin will continue to function fine without the view, at least until there's a better solution (eg not using views :) )

**Comment by pippinsplugins** [comment]
@smerriman this is the first time I've heard of Media Temple not permitting views. Views have been around for a _very_ long time and are in no way non-standard.

Looks like they may be able to give you special permission to create views. See https://mediatemple.net/community/products/grid/204405564/database-privileges-on-the-grid

The view is used on the dashboard stats page and Affiliates > Reports > Campaigns.

**Comment by smerriman** [comment]
That page just says if you need those permissions you have to switch to dedicated hosting, which is a bit of overkill here. 

Yeah, it's not non-standard, but this is still the first plugin I've ever seen that actually uses it, and MediaTemple is a pretty large hosting company. My guess is that all of the users of your plugin on there don't actually realise it's broken currently (due to there being no obvious signs, and a very minor feature).

**Comment by pippinsplugins** [comment]
We've been discussing some graceful fallbacks / alternatives and will see what we can do.

This is the first time we've ever had a report of 100% failure with views due to a hosting environment.

**Comment by sbimarketing** [comment]
Hey guys! @pippinsplugins @smerriman

Actually, I've got another confirmation of a 100% backup failure with views due to the hosting environment!

We use WP Engine for the majority of our hosting and we've just found that the views created by AffiliateWP are preventing them from running their backups also.

The easiest way for us to get around it so that we can run backups is to just drop the views in the database, but obviously we'd rather not do this.

So if you have an update or any ideas on how we could address this, that'd be great thanks.

**Comment by michaelbeil** [comment]
@sbimarketing The `wp_affiliate_wp_campaigns` table is not an actual table, it is a "view". Since the campaigns view is created automatically from the referrals / visits table, you do not need to include that in your backups. The campaigns view will automatically reflect the changes that happen in the referrals / visits tables after restoring from a backup.

A bit more on views:
> In SQL, a view is a virtual table based on the result-set of an SQL statement.

> A view contains rows and columns, just like a real table. The fields in a view are fields from one or more real tables in the database.

**Comment by pippinsplugins** [comment]
@DrewAPicture could you please reach out to WP Engine. It's silly that views aren't respected in their server configs.

**Comment by DrewAPicture** [comment]
@pippinsplugins Yep, I'll send WPEngine support a message, get the conversation started.

**Comment by sbimarketing** [comment]
@michaelbeil Thanks for your input, that makes sense.

@pippinsplugins You're right, it IS silly that views aren't included in WP Engine's server configs in relation to backups.

@DrewAPicture Thanks! Hopefully you can work some magic with WP Engine, because I've brought it up with them before, but the "view issue" probably hasn't happened often enough for them to make it a priority to attend to!

**Comment by pippinsplugins** [comment]
New report of this https://secure.helpscout.net/conversation/360810019/58759/?folderId=205524

I don't know if there's anything we can do here, but let's please take another look @ramiabraham @DrewAPicture

**Comment by pippinsplugins** [comment]
This new report was also with WP Engine.

**Comment by pippinsplugins** [comment]
From [MySQL doc site](https://dev.mysql.com/doc/refman/5.7/en/view-restrictions.html):

>>>There exists a shortcoming with the current implementation of views. If a user is granted the basic privileges necessary to create a view (the CREATE VIEW and SELECT privileges), that user will be unable to call SHOW CREATE VIEW on that object unless the user is also granted the SHOW VIEW privilege.

>>>That shortcoming can lead to problems backing up a database with mysqldump, which may fail due to insufficient privileges. This problem is described in Bug #22062.

>>>The workaround to the problem is for the administrator to manually grant the SHOW VIEW privilege to users who are granted CREATE VIEW, since MySQL doesn't grant it implicitly when views are created.

**Comment by smerriman** [comment]
Just wanted to point out - the original purpose of this ticket was pointing out that some hosts like MediaTemple block views from being *created*, meaning the entire view functionality in the plugin does not work and results in errors.

The fact it causes issues with backups is really a separate issue so perhaps it should be branched off.

**Comment by pippinsplugins** [comment]
We've had a new case of this happening on a GoDaddy site. I've reach out to a tech there.

**Comment by pippinsplugins** [comment]
New case of this on WPEngine https://secure.helpscout.net/conversation/376840712/60625/?folderId=205524

**Comment by planetahuevo** [comment]
For your records, ended up here after having the same issue trying to import the database and having problems with manual backups on Siteground. 
I have not tested yet on the new server after importing without creating the view but I though it was good to say this "view" things still cause some small issues.
